base/common/src/org/dogtagpki/nss/NSSExtensionGenerator.java

```diff
@@ -0,0 +1,452 @@
+import org.mozilla.jss.netscape.security.x509.CPSuri;
+import org.mozilla.jss.netscape.security.x509.CertificateExtensions;
+import org.mozilla.jss.netscape.security.x509.CertificatePoliciesExtension;
+import org.mozilla.jss.netscape.security.x509.CertificatePolicyId;
+import org.mozilla.jss.netscape.security.x509.CertificatePolicyInfo;
+import org.mozilla.jss.netscape.security.x509.GeneralName;
+import org.mozilla.jss.netscape.security.x509.KeyIdentifier;
+import org.mozilla.jss.netscape.security.x509.KeyUsageExtension;
+import org.mozilla.jss.netscape.security.x509.PolicyQualifierInfo;
+import org.mozilla.jss.netscape.security.x509.PolicyQualifiers;
+import org.mozilla.jss.netscape.security.x509.SubjectKeyIdentifierExtension;
+import org.mozilla.jss.netscape.security.x509.URIName;
+import org.mozilla.jss.netscape.security.x509.X509CertImpl;
+import org.mozilla.jss.netscape.security.x509.X509Key;
+
+import com.netscape.cmsutil.crypto.CryptoUtil;
+
+/**
+ * @author Endi S. Dewata
+ */
+public class NSSExtensionGenerator {
+
+    public static org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(NSSExtensionGenerator.class);
+
+    private Map<String, String> parameters = new LinkedHashMap<String, String>();
+
+    public NSSExtensionGenerator() {
+    }
+
+    /**
+     * Initialize cert extension generator with configuration file
+     * based on the following format:
+     * https://www.openssl.org/docs/manmaster/man5/x509v3_config.html
+     */
+    public void init(String filename) throws Exception {
+
+        Properties properties = new Properties();
+        properties.load(new FileInputStream(filename));
+
+        parameters.clear();
+        for (String name : properties.stringPropertyNames()) {
+            String value = properties.getProperty(name);
+            parameters.put(name, value);
+        }
+    }
+
+    public Map<String, String> getParameters() {
+        return parameters;
+    }
+
+    public void setParameters(Map<String, String> parameters) {
+        this.parameters.clear();
+        this.parameters.putAll(parameters);
+    }
+
+    public Collection<String> getParameterNames() {
+        return parameters.keySet();
+    }
+
+    public Collection<String> getParameterNames(String parent) {
+
+        String prefix = parent + ".";
+        int length = prefix.length();
+
+        return parameters.keySet().stream()
+            .filter(name -> name.startsWith(prefix))
+            .map(name -> name.substring(length))
+            .collect(Collectors.toSet());
+    }
+
+    public String getParameter(String name) {
+        return parameters.get(name);
+    }
+
+    public void setParameter(String name, String value) {
+        parameters.put(name, value);
+    }
+
+    public String removeParameter(String name) {
+        return parameters.remove(name);
+    }
+
+    public BasicConstraintsExtension createBasicConstraintsExtension() throws Exception {
```

Comment:
I can guess at usage but it'd be nice to have documentation about how to actually use these parameters :-)

Reply:
For now I can only point to the [OpenSSL doc](https://www.openssl.org/docs/manmaster/man5/x509v3_config.html) which the config format was based on. It's a big doc and we haven't really decided how much to adopt from it. We can build the doc incrementally in separate PRs later.